railties/lib/rails/application/finisher.rb

```diff
@@ -93,6 +94,28 @@ module Finisher
         end
       end
 
+      initializer :notify_early_loaded_component do
+        next if config.eager_load
+
+        rails_components = [
+          :action_view,
+          :action_cable,
+          :action_controller,
+          :active_job,
+          :action_mailer,
+          :active_record
+        ]
+        loaded_components = ActiveSupport.loaded_components.keys & rails_components
+        if loaded_components.any?
+          $stderr.puts(<<~EOM)
+            Your application (or a dependencies) is making use of Rails component too early on boot.
+            This can cause various issues (such as ignoring some configuration values you set) and increase boot time.
+            To discard this warning, use lazy load hooks instead.
+          EOM
+          ActiveSupport::Notifications.instrument("components_loaded.rails", components: loaded_components)
```

Comment:
Do we need a notification given we are already showing a warning on stderr? We could just print the loaded components on the warning. If the notification was when the component was loaded it would be more useful. Right now it is just tell you something is wrong, not not what is wrong and where it is wrong.

Reply:
Yes agree to tweak the logging message and include the component(s) name. I think sending a notification is valuable as I want to have the flexibility to make the application hard fail when it boots

Reply:
But what information the exception will bring? It will happen after the initialize run, not when the component is loaded, so the exception will show the same info as the log message.

Reply:
It won't bring anything else than the log, but at least if a developer writes code and make the mistake or add a new gem that has this issue, we can prevent it. I can definitely people not noticing the issue/ignoring it. The same way it is for deprecations (people ignore it so we had to make tests to fail)